### ISSUE TITLE
`optuna.importance.get_param_importances` sometimes skips params

### ISSUE THREAD

**Issue by PhilipMay** [optuna/optuna#1856]
When doing `optuna.importance.get_param_importances(study)` some features are sometimes skipped.
Example:

![grafik](https://user-images.githubusercontent.com/229382/93984526-3383af80-fd84-11ea-9aff-b68beda33c6a.png)

While `plot_slice` shows this:

![grafik](https://user-images.githubusercontent.com/229382/93984674-63cb4e00-fd84-11ea-8be2-a0fb962ebbf2.png)

`get_param_importances` somehow skipped `learning_rate`. How does this come? Does optuna only display the most important?

**Comment by hvy** [comment]
Hm I see. Do you have the code for reproduction? Conditional hyperparameters will be skipped (unless you specify which parameters to evaluate explicitly using the `params` parameter) but I would guess that `learning_rate` isn't conditioned?

**Comment by PhilipMay** [comment]
Yes. learning_rate is not conditioned.

I switched to this sampler: `sampler = optuna.samplers.CmaEsSampler(restart_strategy='ipop')`

And get this warning:

```
[W 2020-09-23 09:39:55,691] The parameter 'learning_rate' in trial#53 is sampled independently by using `RandomSampler` instead of `CmaEsSampler` (optimization performance may be degraded). You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
```

Is this warning in relation to the problem?

I can not print the full code here because it is business related. I am doing a cross validation and call this in a loop:

``` python
def train_on_split(silo_to_use, trial, max_seq_len, batch_size, class_weights, step, processor):
    dropout = trial.suggest_uniform('dropout', 0.0, 0.5)
    learning_rate = trial.suggest_uniform('learning_rate', 2e-7, 2e-4)  # 3e-5
    epochs = trial.suggest_int('epochs', 1, 5)  # 4
    warmup_proportion = trial.suggest_uniform('warmup_proportion', 0.1, 0.5)  # 0.2
```

**Comment by PhilipMay** [comment]
@hvy is there anything I can help you with to debug this?

**Comment by hvy** [comment]
Thanks for the follow-up. I don't think that falling back to independent sampling should affect the hyperparameter importance results. However, as you mention, they might just be two different manifestations of the same underlying behavior. Could you maybe check if the `learning_rate` parameter is present in all the trials belonging to the study for which you're assessing the importances? You could loop over all trials and check the attributes (`t.params for t in Study.trials`), dump the trials into a dataframe `Study.trials_dataframe()` or query the RDB directly if you are using an `RDBStorage`.

**Comment by PhilipMay** [comment]
All complete trials have LR set. There is one failed and one running trial stored in the DB which has `NaN` stored in LR.

**Comment by hvy** [comment]
Does the following simulate your study?

```python
import optuna

def objective(trial):
    x = trial.suggest_float("x", 0, 1)
    y = trial.suggest_float("y", 0, 1)

    if trial.number == 1:
        raise RuntimeError

    # x = trial.suggest_float("x", 0, 1)
    # y = trial.suggest_float("y", 0, 1)

    return x ** 2 + y

study = optuna.create_study()
study.optimize(objective, n_trials=32, catch=(RuntimeError,))

study_copy = optuna.create_study()
for t in study.trials:
    if t.number == 2:
        t.state = optuna.trial.TrialState.RUNNING
        t.datetime_complete = None
        t.value = None
        t.params = {}
        t.distributions = {}
    study_copy.add_trial(t)

importances = optuna.importance.get_param_importances(study_copy)
assert "x" in importances
assert "y" in importances
print(importances)
```

The thing is that trials that are not completed (`Trial.state != TrialState.COMPLETE`) should simply be ignored when evaluating the importances.

**Comment by PhilipMay** [comment]
@hvy yes. I think that should simulate what I have in my Optuna DB when this issue appears.
But as you said. Everything that is not complete should be ignored. And all completed jobs have a learning rate set.

Could it be that my Optuna DB is "old" and I am using a new Optuna version and there are some incompatibilities?
Although I get no warnings or something.

**Comment by PhilipMay** [comment]
@hvy I exported the DB out of the cloud environment for you:
[optuna_copy.zip](https://github.com/optuna/optuna/files/5297210/optuna_copy.zip)

```python
study_name='copy'
storage='sqlite:///optuna_copy.db'
study = optuna.create_study(study_name=study_name, 
                            storage=storage,
                            load_if_exists=True, direction='maximize')
df = study.trials_dataframe()
print(df)
fig = optuna.visualization.plot_param_importances(study)
fig.write_html("../fig/optuna_plot_importances_{}.html".format(study_name))
```

**Comment by hvy** [comment]
Thanks for sharing your study. I now see that the `low` differs between trials for the `learning_rate` (2e-6 for the first trials,  2e-7 for the later ones). This is the code of interest https://github.com/optuna/optuna/blob/master/optuna/samplers/_search_space.py#L75 where it matters. 

The importance being evaluated (with fANOVA) depends on the search space definitions and since this makes the search space ambiguous for the `learning_rate`, it's also skipped by the algorithm. We might be able to make it more clear to the user but I'm not sure if we can change this logic.

**Comment by PhilipMay** [comment]
Hmm - ok. Thanks for this analysis. For me it is a very normal situation that I open up the search space during the search. 

Maybe you could think about a way on how to adapt this.

PS: Do you think the documentation should give a hint about this precondition? I could do a PR for this.

What do you think?

**Comment by hvy** [comment]
> PS: Do you think the documentation should give a hint about this precondition? I could do a PR for this.

Yes, that'd be great! Going through the logic in [`_get_distributions`](https://github.com/optuna/optuna/blob/master/optuna/importance/_base.py#L49-L91) might be a good start for that. Note that the logic differs depending on whether parameters (via the `params` argument to `get_param_importances`) are specified (in this case your example should yield a `ValueError`).

> Maybe you could think about a way on how to adapt this.

Actually, we might be able to use the largest search spaces for those parameters (in this case where `low` of `learning_rate` is 2e-7). In contrast to the samplers where this might hurt the algorithm, the importance evaluator would just think that there were few samples at these certain regions. @sile, do you happen to have any insights on this?

**Comment by sile** [comment]
> we might be able to use the largest search spaces for those parameters 

@hvy I think that fANOVA can handle this case with no problems.

**Comment by hvy** [comment]
@PhilipMay do you have any updates on the documentation PR? It could just be that I didn't communicate well enough, the next actions to be taken. I'd also like to label this issue as contribution welcome to discuss fixing the logic.

**Comment by PhilipMay** [comment]
> > PS: Do you think the documentation should give a hint about this precondition? I could do a PR for this.
> 
> Yes, that'd be great! Going through the logic in [`_get_distributions`](https://github.com/optuna/optuna/blob/master/optuna/importance/_base.py#L49-L91) might be a good start for that. Note that the logic differs depending on whether parameters (via the `params` argument to `get_param_importances`) are specified (in this case your example should yield a `ValueError`).

Well - I do not have the time to go though code and write documentation about my observations. Sorry.

**Comment by hvy** [comment]
Got it, thanks for the heads up.

Let me label this issue as contribution welcome, preferably fixing the logic.

**Comment by hvy** [comment]
PR is ongoing at #2003.

**Comment by g-votte** [comment]
Since #2003 has been closed, this issue is still open.

**Comment by g-votte** [comment]
Since this issue has been open for a long time, let me close once. Feel free to reopen upon necessity.

**Comment by PhilipMay** [comment]
IMO this is an issue of / in Optuna.
Do you really want to just close it?

**Comment by g-votte** [comment]
Sorry that I didn't explain enough. The action above is just to filter out many other out-dated issues.

If the problem is still reproduced, we should reopen.

**Comment by PhilipMay** [comment]
> Sorry that I didn't explain enough. The action above is just to filter out many other out-dated issues.
> 
> If the problem is still reproduced, we should reopen.

So you say the issue might be fixed and I should do a regression test?

**Comment by hvy** [comment]
I think it's still an issue in master https://github.com/optuna/optuna/commit/45b0a1c422169f34671066c7a06495756e2fd4c8. It's been a while so please correct me if I'm wrong but this was the code I tested.

```python
import optuna as ot


def objective(t):
    x = t.suggest_float("x", 0, 1)
    y = t.suggest_float("y", x, 1)  # Lower bound is intentionally dynamic.
    return x + y


if __name__ == "__main__":
    study = ot.create_study()
    study.optimize(objective, n_trials=100)
    imps = ot.importance.get_param_importances(study)
    assert "x" in imps  # Ok.
    assert "y" in imps  # Fails, since y's lower bound is dynamic.
```

Perhaps we can reopen it. It's also been mentioned in #3889.

**Comment by g-votte** [comment]
Thanks @PhilipMay for reporting again and @hvy for analysis!
As the problem still reproduces, let's reopen.

**Comment by PhilipMay** [comment]
> This issue has not seen any recent activity.

But please do not close it. :-)

**Comment by sgrubas** [comment]
Hello,

I am currently having the same issue of skipping params with dynamic range (the code below throws an AssertionError)

Any plans or updates on it?

> I think it's still an issue in master [45b0a1c](https://github.com/optuna/optuna/commit/45b0a1c422169f34671066c7a06495756e2fd4c8). It's been a while so please correct me if I'm wrong but this was the code I tested.
> 
> ```python
> import optuna as ot
> 
> 
> def objective(t):
>     x = t.suggest_float("x", 0, 1)
>     y = t.suggest_float("y", x, 1)  # Lower bound is intentionally dynamic.
>     return x + y
> 
> 
> if __name__ == "__main__":
>     study = ot.create_study()
>     study.optimize(objective, n_trials=100)
>     imps = ot.importance.get_param_importances(study)
>     assert "x" in imps  # Ok.
>     assert "y" in imps  # Fails, since y's lower bound is dynamic.
> ```
> 
> Perhaps we can reopen it. It's also been mentioned in #3889.

**Comment by nzw0301** [comment]
Since this issue has been closed, no plan to resolve this issue, I think.
